### DISCUSSION TITLE
Can HexBlock.verifyDims be generalized or exposed to plugins?

### DISCUSSION THREAD

**Discussion by drewj-usnctech** [terrapower/armi#640] [Ideas]
The current implementation of `HexBlock` feels heavily tied towards sodium fast reactor design with some baked-in assumptions

- tightly packed fuel pins
- separated by wire wrap
- all the way out to the duct

https://github.com/terrapower/armi/blob/b9eb9cae4044405dc01a11fea323d5dd44f862d2/armi/reactor/blocks.py#L1745-L1757

Any reactor not adhering to this design is likely to see a lot of warnings in their runlog like
```
[info] Constructing assembly `fuel`
[warn] No component matched Flags.WIRE in <fuel block-bol-000 at ... >. Returning None
[warn] No component matched Flags.CLAD in <fuel block-bol-000 at ... >. Returning None
```

This feels like something where generalization / plugins could be useful. But it both needs to be more general but also design specific. Exposing this to a plugin that handles sodium fast reactors _and_ LWRs would need some additional logic to know how to check the validity of a `Block`

There's also this section of code that will never be executed because it's directly after an exception in the same scope

https://github.com/terrapower/armi/blob/b9eb9cae4044405dc01a11fea323d5dd44f862d2/armi/reactor/blocks.py#L1774-L1786

**Comment by keckler** [comment]
I very much agree that `HexBlock` should not be so specific. Even for SFRs, it is overly specific. 

So there is that.

But I also think that you should be able to make your own block type from within a plugin that is recognized when building a case, without any changes needed to the current framework. I might be wrong about that, but I know it is possible for `components` and I anticipate it could work the same for `blocks`. 

Is this what you are hoping to do? Maybe I am unclear about how your thoughts on how to potentially resolve this.

**Reply by drewj-usnctech** [reply]
@keckler Hmm I don't think having the `Block`-type be potentially reactor specific is what I was envisioning. Can you expand more on

> But I also think that you should be able to make your own block type from within a plugin that is recognized when building a case, without any changes needed to the current framework

as I may be missing something.

I think more what I was looking for would be a solution where a `Plugin` can perform the validation on a block. So something like a SFR Block validator would look for
- number of pins fully fills a hexagonal ring structure
- wire and clad to duct ratio item from the snippet above

but then applications could build, e.g., an FFTF-validator, something for the MHTGR (#224), C5G7 validator, etc.

**Reply by keckler** [reply]
Right, so you are definitely thinking more broadly than I was. And your solution might be better in the long run, but I'm imagining it will take more lifting up-front.

I was basically just thinking that you could implement your own block class from within your plugin. It could essentially just be a direct subclass of the `HexBlock` class with its own implementation of the `verifyBlockDims` method. And then you could use this `MySubclassedHexBlock` or whatever in building your inputs, and not have to deal with the stupid checks that are currently implemented on `HexBlock.verifyBlockDims`.

I just didn't know if you knew that it was a possibility to define your own blocks in your plugins. If you did know that already, then sorry for the confusion! And if you _didn't_ know that and want me to help explain that, I'd be happy to try. Just let me know.

**Reply by drewj-usnctech** [reply]
Ah gotcha. I just found these two plugin hooks that look like they could work

https://github.com/terrapower/armi/blob/eeb28b46de305cf2477ca92fd2aec811704fdd13/armi/plugins.py#L253-L286

now things make more sense.

Thanks!
